### PR TITLE
Implement basic (recommended against) support for inheriting from COM interfaces defined in different assemblies

### LINKED ISSUE #102602: compilation gives error SYSLIB1094 for a GeneratedComInterface interface deriving from an interface in another project
### Description

I have two projects, one console app project, referencing another class library project. One interface `IMy` is declared in the class library, and another interface `IMy3` is declared in the console app and derives from `IMy`

Both projects are using ComWrappers source generation feature (`GeneratedComInterface`).

I cannot compile the console application.

### Reproduction Steps

Here the class library project:

```
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<TargetFramework>net8.0</TargetFramework>
		<ImplicitUsings>enable</ImplicitUsings>
		<Nullable>enable</Nullable>
		<AllowUnsafeBlocks>True</AllowUnsafeBlocks>
		<PublishAot>true</PublishAot>
	</PropertyGroup>
</Project>

```
Contains only one file IMy.cs:

```
using System.Runtime.InteropServices;
using System.Runtime.InteropServices.Marshalling;

namespace AotTest
{
    [GeneratedComInterface, Guid("68207504-314f-492a-b4f2-fbc1838e6ac3")]
    public partial interface IMy
    {
    }

    [GeneratedComInterface, Guid("80511349-dff5-4c5e-839b-af4735977ebf")]
    public partial interface IMy2 : IMy // this derived one works fine
    {
    }
}
```

Here's the console app project: 

```
<Project Sdk="Microsoft.NET.Sdk">
	<PropertyGroup>
		<OutputType>Exe</OutputType>
		<TargetFramework>net8.0</TargetFramework>
		<ImplicitUsings>enable</ImplicitUsings>
		<Nullable>enable</Nullable>
		<PublishAot>true</PublishAot>
		<AllowUnsafeBlocks>True</AllowUnsafeBlocks>
		<InvariantGlobalization>true</InvariantGlobalization>
	</PropertyGroup>

	<ItemGroup>
		<ProjectReference Include="..\AotTest\AotTest.csproj" />
	</ItemGroup>
</Project>

```
contains only one file Program.cs:

```
using System.Runtime.InteropServices;
using System.Runtime.InteropServices.Marshalling;

namespace AotTest.Cli
{
    internal class Program
    {
        static void Main()
        {
            Console.WriteLine("Hello, World!");
        }
    }

    [GeneratedComInterface, Guid("755293e8-fe80-4654-aee3-732fefaaf3e0")]
    public partial interface IMy3 : IMy
    {
    }
}

```

### Expected behavior

I don't see why it shouldn't compile, I hope this simple scenario is supported.

### Actual behavior

I get compilation error:

```
1>------ Build started: Project: AotTest, Configuration: Debug Any CPU ------
1>AotTest -> E:\smo\work\AotTest\AotTest\bin\Debug\net8.0\AotTest.dll
2>------ Build started: Project: AotTest.Cli, Configuration: Debug Any CPU ------
2>E:\smo\work\AotTest\AotTest.Cli\Program.cs(15,30,15,34): error SYSLIB1094: COM interface AotTest.Cli.IMy3 inherits from AotTest.IMy, which has errors. ComInterfaceGenerator will not generate source for AotTest.Cli.IMy3. (https://learn.microsoft.com/dotnet/fundamentals/syslib-diagnostics/syslib1094)
2>Done building project "AotTest.Cli.csproj" -- FAILED.
========== Build: 1 succeeded, 1 failed, 0 up-to-date, 0 skipped ==========
```

Also, independently from the SYSLIB1094 error, Roslyn seems to crash (but not always, seems more a background thing) this is what I see (sometimes) in Visual Studio and it seems somehow related (I have a bigger project where these top error lines popup continuously):

image

Here is the stack trace from the first error line:
```
StreamJsonRpc.RemoteInvocationException: SyntaxTree is not part of the compilation (Parameter 'syntaxTree')
   at StreamJsonRpc.JsonRpc.<InvokeCoreAsync>d__154`1.MoveNext()
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.CodeAnalysis.Remote.BrokeredServiceConnection`1.<TryInvokeAsync>d__20`1.MoveNext()
RPC server exception:
System.ArgumentException: SyntaxTree is not part of the compilation (Parameter 'syntaxTree')
      at Microsoft.CodeAnalysis.CSharp.CSharpCompilation.GetSemanticModel(SyntaxTree syntaxTree, SemanticModelOptions options)
      at Microsoft.CodeAnalysis.Diagnostics.SuppressMessageAttributeState.IsDiagnosticSuppressed(Diagnostic diagnostic, SuppressMessageInfo& info)
      at Microsoft.CodeAnalysis.Diagnostics.SuppressMessageAttributeState.ApplySourceSuppressions(Diagnostic diagnostic)
      at Microsoft.CodeAnalysis.GeneratorDriver.FilterDiagnostics(Compilation compilation, ImmutableArray`1 generatorDiagnostics, DiagnosticBag driverDiagnostics, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.GeneratorDriver.RunGeneratorsCore(Compilation compilation, DiagnosticBag diagnosticsBag, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.GeneratorDriver.RunGenerators(Compilation compilation, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.SolutionCompilationState.CompilationTracker.ComputeNewGeneratorInfoInCurrentProcessAsync(ISourceGeneratorTelemetryCollectorWorkspaceService telemetryCollector, Compilation compilationWithoutGeneratedFiles, TextDocumentStates`1 oldGeneratedDocuments, GeneratorDriver generatorDriver, Compilation compilationWithStaleGeneratedTrees, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.SolutionCompilationState.CompilationTracker.ComputeNewGeneratorInfoAsync(SolutionCompilationState compilationState, Compilation compilationWithoutGeneratedFiles, TextDocumentStates`1 oldGeneratedDocuments, GeneratorDriver oldGeneratorDriver, Compilation compilationWithStaleGeneratedTrees, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.SolutionCompilationState.CompilationTracker.AddExistingOrComputeNewGeneratorInfoAsync(CreationPolicy creationPolicy, SolutionCompilationState compilationState, Compilation compilationWithoutGeneratedFiles, CompilationTrackerGeneratorInfo generatorInfo, Compilation compilationWithStaleGeneratedTrees, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.SolutionCompilationState.CompilationTracker.<>c__DisplayClass23_0.<<GetOrBuildFinalStateAsync>g__FinalizeCompilationWorkerAsync|4>d.MoveNext()
   --- End of stack trace from previous location ---
      at Microsoft.CodeAnalysis.SolutionCompilationState.CompilationTracker.<>c__DisplayClass23_0.<<GetOrBuildFinalStateAsync>g__FinalizeCompilationAsync|3>d.MoveNext()
   --- End of stack trace from previous location ---
      at Microsoft.CodeAnalysis.SolutionCompilationState.CompilationTracker.<>c__DisplayClass23_0.<<GetOrBuildFinalStateAsync>g__BuildFinalStateAsync|0>d.MoveNext()
   --- End of stack trace from previous location ---
      at Microsoft.CodeAnalysis.SolutionCompilationState.CompilationTracker.GetOrBuildFinalStateAsync(SolutionCompilationState compilationState, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.SolutionCompilationState.CompilationTracker.GetSourceGeneratedDocumentStatesAsync(SolutionCompilationState compilationState, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.Remote.RemoteSourceGenerationService.<>c__DisplayClass2_0.<<GetSourceGenerationInfoAsync>b__0>d.MoveNext()
   --- End of stack trace from previous location ---
      at Microsoft.CodeAnalysis.Remote.RemoteWorkspace.<>c__DisplayClass8_0`1.<<RunWithSolutionAsync>g__ProcessSolutionAsync|1>d.MoveNext()
   --- End of stack trace from previous location ---
      at Microsoft.CodeAnalysis.Remote.RemoteWorkspace.RunWithSolutionAsync[T](AssetProvider assetProvider, Checksum solutionChecksum, Int32 workspaceVersion, Boolean updatePrimaryBranch, Func`2 implementation, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.Remote.RemoteWorkspace.RunWithSolutionAsync[T](AssetProvider assetProvider, Checksum solutionChecksum, Int32 workspaceVersion, Boolean updatePrimaryBranch, Func`2 implementation, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.Remote.BrokeredServiceBase.RunWithSolutionAsync[T](Checksum solutionChecksum, Func`2 implementation, CancellationToken cancellationToken)
      at Microsoft.CodeAnalysis.Remote.BrokeredServiceBase.RunServiceImplAsync[T](Func`2 implementation, CancellationToken cancellationToken)

```

### Regression?

I don't tink so but I can't be sure.

### Known Workarounds

I don't have any.

### Configuration

.NET 8.0.5, Visual Studio 17.10.0 (latest), Windows 11 x64

Here's a reproducing solution
[AotTest.zip](https://github.com/dotnet/runtime/files/15415771/AotTest.zip)

### Other information

_No response_

### PR DESCRIPTION
**Pull request by jkoritzinsky**
Fixes #102602

This PR provides support for simple scenarios of interface inheritance across different assemblies. Using this feature emits a user-suppressible warning with pointers to our (to be updated) documentation.

This feature may not work in advanced scenarios. We expect that the majority of usage of the COM source generator will not use this feature, so that is okay. The following are some scenarios that may not work today or in the future:

- The base interface type defines a method that shadows a method in its base interface.
- The base interface type is part of a project compiled using ComInterfaceGenerator vX, which has one set of rules for how methods are defined on a COM interface. The derived type is compiled using ComInterfaceGenerator vY, which has another set of rules (property support in the future comes to mind).
- The base interface's project had various source generator errors suppressed, and the base interface is defined incorrectly.

Additionally, having a COM interface inheritance chain defined across assemblies causes problems with assembly versioning and creates a really bad pit of failure if the author violates the rules of COM and changes the vtable of their COM interface type (see https://github.com/dotnet/runtime/issues/102602#issuecomment-2223957741).

We believe that for the amount of usage we expect for this feature, providing an effectively "opt-in and agree to the pitfalls" experience provides the necessary guardrails to avoid guiding users into a pit of failure.

### CODE REVIEW

**Comment by jkoritzinsky** [comment]
@dotnet/appmodel PTAL

File: src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceInfo.cs [hunk 1]
```diff
@@ -101,7 +105,67 @@ public static DiagnosticOrInterfaceInfo From(INamedTypeSymbol symbol, InterfaceD
...
                     interfaceAttributeData.Options,
                     syntax.Identifier.GetLocation()),
-                symbol));
+                symbol);
+
+            // Now that we've validated all of our requirements, we will check for some non-blocking scenarios
+            // and emit diagnostics.
+            ImmutableArray<DiagnosticInfo>.Builder nonFatalDiagnostics = ImmutableArray.CreateBuilder<DiagnosticInfo>();
+
+            if (baseSymbol is not null
```
Comment (AaronRobinsonMSFT) [inline src/libraries/System.Runtime.InteropServices/gen/ComInterfaceGenerator/ComInterfaceInfo.cs:114]: Can we comment what this is doing. It isn't clear from reading it, but I think the point is the base type is not one we have from source or it is from source, but the compilation is not the one we're working on? That last preciate is a bit dense.

**Review by AaronRobinsonMSFT — APPROVED** [review]

**Review by elinor-fung — APPROVED** [review]

File: src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/TargetSignatureTests.cs [hunk 2]
```diff
@@ -296,19 +297,101 @@ partial interface IComInterface
...
+                using System.Runtime.InteropServices;
+                using System.Runtime.InteropServices.Marshalling;
+                
+                [GeneratedComInterface]
+                [Guid("0A617667-4961-4F90-B74F-6DC368E9817A")]
+                partial interface {|#1:IComInterface2|} : IComInterface
+                {
+                    void DerivedMethod();
+                }
+                """;
```
[the remark this answers is not in the document]
Comment (jkoritzinsky) [reply src/libraries/System.Runtime.InteropServices/tests/ComInterfaceGenerator.Unit.Tests/TargetSignatureTests.cs:327]: I'll add some more tests for this case.

**Pull request merged**